src/library/scala/concurrent/Future.scala

```diff
@@ -485,9 +495,9 @@ trait Future[+T] extends Awaitable[T] {
   def andThen[U](pf: PartialFunction[Try[T], U])(implicit executor: ExecutionContext): Future[T] =
     transform {
       result =>
-        try pf.applyOrElse[Try[T], Any](result, Predef.identity[Try[T]])
-        catch { case NonFatal(t) => executor reportFailure t }
-
+        try pf.applyOrElse[Try[T], Any](result, Future.id[Try[T]])
+        catch { case t if NonFatal(t) => executor.reportFailure(t) }
```

Comment:
It's more verbose but more efficient than the pattern matching:

```scala
catch {
  case ex =>
    if(NonFatal(ex))  executor.reportFailure(t) 
    else throw ex
}
```

Reply:
@fwbrasil Interesting. Will have a look!